### DISCUSSION TITLE
FLIP transitions

### DISCUSSION THREAD

**Discussion by Ahrengot** [livewire/livewire#7478] [3. Feature ideas]
This idea relates to

- https://github.com/livewire/livewire/discussions/4133
- https://github.com/livewire/livewire/discussions/4702

## Concept
![flip-animation](https://github.com/livewire/livewire/assets/391810/446b02de-c865-40c0-91ff-64247c5ae8ec)

Having built-in support for [flip animations](https://css-tricks.com/animating-layouts-with-the-flip-technique/) would make it very easy to nail that SPA-feeling in any scenario where you are adding, removing, filtering and sorting elements. The current implementation of `wire:transition` works well for toggling a piece of UI on and off, but it falls short when dealing with anything more complicated.

While it's fairly easy to roll your own flip animation solution and add it as an alpine directive, I think it would be good to have a standard solution for Livewire that everybody can rely on and make improvements to.

I also think it makes sense to have a seperate directive for flip-animations, since they do differ from normal transitions. Maybe something like `<div wire:flip>`

## Approach
It would make sense to add this as a plugin or directive in Alpine.js and add a livewire helper, sort of how the current `wire:transition` implementation works. It's "just" UI-behaviour, so nothing would tie it directly to the Livewire lifecycle as far as I can tell.

## Inspiration
In terms of the animation logic itself, I think [AutoAnimate](https://auto-animate.formkit.com/) is a good place to look for inspiration. They have a pretty straight-forward implementation that's really easy to work with.

**Comment by Ahrengot** [comment]
I wrote a basic implementation using the autoanimate library as a proof of concept, and it seems to work just fine with Livewire.

alpine-flip.js
```JS
import autoAnimate from '@formkit/auto-animate';
import {Alpine} from '...';

document.addEventListener("alpine:init", () => {
    Alpine.data("flip", () => ({
        controller: null,
        init() {
            this.controller = autoAnimate(this.$el);
        },
        destroy() {
            this.controller.destroy();
        }
    }));
});
```

some.blade.php
```HTML
<div x-data="flip" class="space-y-3">
    @forelse($items as $item)
        <div wire:key="{{ $item->id }}">
            ...
        </div>
    @empty
        <x-empty-state />
    @endforelse
</div>
```

**Reply by DanielSpravtsev** [reply]
Thanks for your snippet. It is the feature I alwayse missing for, but don't think that it can be added to the core because it's 3rd party. Would be awesome to pack in into a plugin

**Reply by Ahrengot** [reply]
Agreed. I don't think the auto-animate package should be a dependency for Alpine/Livewire. Just used it as an example here. 

Just edited the post, as I can see how it could be read as if I'm suggesting adding another library as a dependency to Alpine.
